### ISSUE TITLE
updating existing model 

### ISSUE THREAD

**Issue by roypyp** [MaartenGr/BERTopic#498]
Hi Maarten,
we currently have a working model with a produced list of topics, but we want the model to be able when given new documents to update the existing topics automatically or create new topics if the documents don't fit the existing topics.
we are struggling to do so and were wondering if there is a way to do so?

Thanks in advance and have a great day!

**Comment by MaartenGr** [comment]
This is currently not possible in BERTopic and definitely not a trivial thing to achieve. It would mean that UMAP and HDBSCAN should be updated to accommodate these new points and the predictions should be shifted towards those new points, which changes the underlying predictive nature of these models. 

Typically, I would advise re-training BERTopic if you have a significant amount of new data coming in. Then, to track topics between models, you can use the c-TF-IDF matrix to calculate the similarity between topics in order to match them. Moreover, you can use the [semi-supervised method](https://maartengr.github.io/BERTopic/getting_started/supervised/supervised.html) to improve re-training the model periodically. This method allows you to steer the training towards specific labels that you have. So, whenever you train a BERTopic models, your labels are the topics that were being generated. Then, whenever new data comes in, you combine that data with the data that you already have on train a new BERTopic model on that. However, this time you assign the data with labels, namely the topics that you had already generated before. The new data gets the label -1 as you do not know which topics they will end up in. That way, you keep the general structure of the existing topics and allow new documents to generate new topics. 

As mentioned, to track the topics that you generated in the first training with every new training instance, you can calculate the cosine similarities between the two c-TF-IDF matrices and find the topics that are most similar. 

Hopefully, this helps a bit!

**Comment by dilaraadem** [comment]
@MaartenGr about updating a model, is it possible to reduce the number of topics after saving the trained model and later on calling it with "load" function? I can visualize and access list of topics easily, but since the there are too many topics generated, I want to reduce number of topics possibly without training the model again. When I try to reduce topics, I am getting an error:
![image](https://user-images.githubusercontent.com/32614552/163056860-dbbf337e-8628-4953-af9f-99061a2191d4.png)
Can you tell me if what I am trying to do is possible or not?

**Comment by MaartenGr** [comment]
@dilaraadem From what I can see in your error message, the `loaded_model.topics` is not the variable that you want to pass. You will need to pass a list of topics where each element is the topic related to the document. In other words, you will need to pass `topics` instead of `loaded_model.topics`. The `topics` variable is typically generated as follows:

```python
topic_model = BERTopic()
topics, probs = topic_model.fit_transform(docs)
```

**Comment by dilaraadem** [comment]
@MaartenGr Thanks for the quick answer! 
I am loading the model from a saved version as `loaded_model = BERTopic.load("models/model_name.bt")`. Therefore I don't have the topics or probabilities saved as variable at that runtime instance. If I use the code snippet you wrote above, it means I would have to retrain the model with my data, which I already did before and saved the model in a folder by `topic_model.save`. And I don't want to repeat that step if possible. Is there a way that I can get the topics variable without using `fit_transform` method?

**Comment by MaartenGr** [comment]
@dilaraadem In that case, you can try the following and see if that works:

```python
probs = topic_model.hdbscan_model.probabilities_
topics = topic_model._map_predictions(topic_model.hdbscan_model.labels_)
```

**Comment by roypyp** [comment]
@MaartenGr sorry for the late response 
we tried to train the semi-supervised model approach. 
but after we got the new topics we noticed that the topic identifier did not correlate with the previous topic.
example: 
first model topic = [1,2,3] 
semi-supervised model = [1s,2s,3s,4s,5s]
1!=1s.
there we used the c_tf_idf to calculate the similarity between topics.
but, the vectors aren't of the same size
Do you still think it is doable?
because the size of the terms is changing when adding new documents.

thanks for the fast response.

**Comment by MaartenGr** [comment]
> 1!=1s.

Yes, this is to be expected as the semi-supervised approach is a general technique that also allows you to put in classes, not necessarily topics. Therefore, the resulting topics will not be matched. 

> there we used the c_tf_idf to calculate the similarity between topics.
but, the vectors aren't of the same size
Do you still think it is doable?
because the size of the terms is changing when adding new documents.

It is possible to match them using the `c-TF-IDF` matrix if you make sure the vocabularies match between the matrices. After training your `first_topic_model`, you can access the `CountVectorizer` with `first_topic_model.vectorizer_model.vocabulary_` and pass that to the `semi_supervised_model`:

```python
# Extract vocab from first model
first_topic_model = BERTopic().fit(docs)
vocab = first_topic_model.vectorizer_model.vocabulary_

# Pass that to the second model
from sklearn.feature_extraction.text import CountVectorizer
vectorizer_model = CountVectorizer(vocabulary=vocab)
semi_supervised_model= BERTopic(vectorizer_model=vectorizer_model)
topics, probs = semi_supervised_model.fit_transform(docs, y=y)
```

You could also compare `first_topic_model.topic_embeddings` with `semi_supervised_model.topic_embeddings` using cosine similarity but the c-TF-IDF approach is usually a bit more accurate.

**Comment by roypyp** [comment]
@MaartenGr again sorry for the late response.
we tried to use the topic_embeddings and it gave as good results thank you very much for the help!

**Comment by MaartenGr** [comment]
@nfsedaghat Could you share your code for doing this comparison and creating those models? It is difficult to see what is happening here without seeing the code in action. Moreover, it also depends on the models that you are going to be using. For example, if you are using different UMAP models or if you have not set a random state, then differences are to be expected. Having said that, it would be interesting to see what would happen if you change -1 to a non-negative integer as if it were a single cluster.

**Comment by j-nanda** [comment]
Hi there, I'm trying to use Berertopic even though it is installed, I wonder have any updates been made. i am running it on an Azure compute because when I tried using it like normal on my computer, the kernel kept dying, likely because my excel file with the data that I'm trying to model is too large! Thanks!topic in a jupyter notebook on a compute on Azure learning studio but whenever I try to run the code, it doesn't recognize bertopic even though it is installed, I wonder have any updates been made. i am running it on an Azure compute because when I tried using it like normal on my computer, the kernel kept dying, likely because my excel file with the data that I'm trying to model is too large! Thanks!
![Screenshot 2023-04-13 at 10 42 36 AM (2)](https://user-images.githubusercontent.com/89845704/231797558-83e3a604-cab8-46d5-a310-fe2cb843773c.png)

**Comment by MaartenGr** [comment]
@j-nanda This likely means that BERTopic actually was not installed in the kernel that you are currently using. I believe it would be worthwhile to check whether the specific jupyter kernel that you are using in your notebook has BERTopic installed. If not, install it. If it seems like it is installed, I would advise creating a completely different and clean environment, installing BERTopic there, and then adding the environment as a kernel to jupyter.
